src/schedule/zephyr_dp_schedule_native.c

```diff
@@ -0,0 +1,572 @@
+	task_memory->drv.adapter_ops = &task_memory->ops;
+	mod->dev->drv = &task_memory->drv;
+
+	/* allocate stack - must be aligned and cached so a separate alloc */
+	p_stack = k_thread_stack_alloc(stack_size, options & K_USER);
+	if (!p_stack) {
+		tr_err(&dp_tr, "stack alloc failed");
+		ret = -ENOMEM;
+		goto e_tmem;
+	}
+
+	struct task *ptask = &task_memory->task;
+
+	/* internal SOF task init */
+	ret = schedule_task_init(ptask, uid, SOF_SCHEDULE_DP, 0, ops->run, mod, core, options);
+	if (ret < 0) {
+		tr_err(&dp_tr, "schedule_task_init failed");
+		goto e_stack;
+	}
+
+	struct task_dp_pdata *pdata = &task_memory->pdata;
+
+	/* Point to ksem semaphore for kernel threads synchronization */
+	/* It will be overwritten for K_USER threads to dynamic ones.  */
+	pdata->sem = &pdata->sem_struct;
+	pdata->thread = &pdata->thread_struct;
+
+#ifdef CONFIG_USERSPACE
+	if (options & K_USER) {
+		pdata->sem = k_object_alloc(K_OBJ_SEM);
+		if (!pdata->sem) {
+			tr_err(&dp_tr, "Semaphore object allocation failed");
+			ret = -ENOMEM;
+			goto e_stack;
+		}
+
+		pdata->thread = k_object_alloc(K_OBJ_THREAD);
+		if (!pdata->thread) {
+			tr_err(&dp_tr, "Thread object allocation failed");
+			ret = -ENOMEM;
+			goto e_kobj;
+		}
+		memset(&pdata->thread->arch, 0, sizeof(pdata->thread->arch));
+	}
+#endif /* CONFIG_USERSPACE */
+
+	/* success, fill the structures */
+	pdata->p_stack = p_stack;
+	pdata->mod = mod;
+
+	/* initialize other task structures */
+	ptask->ops.complete = ops->complete;
+	ptask->ops.get_deadline = ops->get_deadline;
+	ptask->priv_data = pdata;
+	*task = ptask;
+
+	/* create a zephyr thread for the task */
+	pdata->thread_id = k_thread_create(pdata->thread, p_stack,
+					   stack_size, dp_thread_fn, ptask, NULL, NULL,
+					   CONFIG_DP_THREAD_PRIORITY, ptask->flags, K_FOREVER);
+
+	/* pin the thread to specific core */
+	ret = k_thread_cpu_pin(pdata->thread_id, core);
+	if (ret < 0) {
+		tr_err(&dp_tr, "zephyr task pin to core failed");
+		goto e_thread;
+	}
+
+#if CONFIG_USERSPACE
+	k_thread_access_grant(pdata->thread_id, pdata->sem, &dp_sync[core]);
+	scheduler_dp_grant(pdata->thread_id, core);
+
+	int pidx;
+	size_t size;
+	uintptr_t start;
+	struct k_mem_partition *ppart[SOF_DP_PART_TYPE_COUNT];
+
+	for (pidx = 0; pidx < ARRAY_SIZE(ppart); pidx++)
+		ppart[pidx] = pdata->mpart + pidx;
+
+	/* Module heap partition */
+	mod_heap_info(mod, &size, &start);
+	pdata->mpart[SOF_DP_PART_HEAP] = (struct k_mem_partition){
+		.start = start,
+		.size = size,
+		.attr = K_MEM_PARTITION_P_RW_U_RW,
+	};
+	/* IPC flattening buffer partitioin */
+	pdata->mpart[SOF_DP_PART_IPC] = (struct k_mem_partition){
+		.start = (uintptr_t)&ipc_buf,
+		.size = sizeof(ipc_buf),
+		.attr = K_MEM_PARTITION_P_RW_U_RW,
+	};
+	/* Host mailbox partition for additional IPC parameters: read-only */
+	pdata->mpart[SOF_DP_PART_CFG] = (struct k_mem_partition){
+		.start = (uintptr_t)MAILBOX_HOSTBOX_BASE,
+		.size = 4096,
+		.attr = K_MEM_PARTITION_P_RO_U_RO,
+	};
```

Comment:
I wonder if permanent access to the IPC memory window is safe. Would this not let threads on the secondary core read and alter every IPC request?

Reply:
Yeah I think we need to copy or at least have a method to copy the IPC payloads. These are not on performance critical path after all.